backend/internal/system/crypto/cryptoservice.go

```diff
@@ -0,0 +1,152 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+// Package crypto provides cryptographic functionality with algorithm agility.
+package crypto
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"sync"
+
+	"github.com/asgardeo/thunder/internal/system/config"
+	"github.com/asgardeo/thunder/internal/system/log"
+	"github.com/asgardeo/thunder/internal/system/utils"
+)
+
+// Algorithm represents supported encryption algorithms
+type Algorithm string
+
+const (
+	// AESGCM represents AES-GCM algorithm
+	AESGCM Algorithm = "AES-GCM"
+	// DefaultKeySize defines the default key size for AES-GCM
+	DefaultKeySize = 32
+)
+
+// CryptoService provides cryptographic operations
+type CryptoService struct {
+	key   []byte
+	keyID string
+	algo  CryptoAlgorithm
+	mu    sync.RWMutex // For thread-safe key updates
+}
+
+var (
+	// instance is the singleton instance of CryptoService
+	instance *CryptoService
+	// once ensures the singleton is initialized only once
+	once sync.Once
+)
+
+// GetCryptoService creates and returns a singleton instance of the CryptoService.
+func GetCryptoService() *CryptoService {
+	logger := log.GetLogger().With(log.String(log.LoggerKeyComponentName, "CryptoService"))
+	once.Do(func() {
+		var err error
+		instance, err = initCryptoService()
+		if err != nil {
+			logger.Error("Failed to initialize CryptoService: %v", log.Error(err))
+		}
+	})
+	return instance
+}
+
+// initCryptoService initializes the CryptoService from configuration sources
+func initCryptoService() (*CryptoService, error) {
+	logger := log.GetLogger().With(log.String(log.LoggerKeyComponentName, "CryptoService"))
+	// Try to get key from the application configuration
+	config := config.GetThunderRuntime().Config.Crypto.Key // Use the correct config getter
+
+	// Check if crypto configuration exists
+	if config != "" {
+		key, err := hex.DecodeString(config)
+		if err == nil {
+			logger.Debug("Using crypto key from configuration")
+			return NewCryptoService(key)
+		}
+		logger.Warn("Invalid crypto key in configuration, generating a new key")
+	}
+
+	// Generate new key as fallback for development
+	logger.Warn("No valid crypto key found in configuration, generating a new one")
+
+	key, err := GenerateRandomKey(DefaultKeySize)
+	if err != nil {
+		return nil, err
+	}
+
+	// Print the generated key for development purposes
+	encodedKey := hex.EncodeToString(key)
+	logger.Debug("Generated new crypto key (hex): %s", log.String("logKey", encodedKey))
+
+	return NewCryptoService(key)
+}
+
+// NewCryptoService creates a new instance of CryptoService with the provided key
+func NewCryptoService(key []byte) (*CryptoService, error) {
+	// Check key size for algorithm
+
+	return &CryptoService{
+		key:   key,
+		keyID: utils.GenerateUUID(), // Unique identifier for the key
```

Comment:
Either the keyId should be derivable from key, or it should be taken as a user input

Reply:
+1. lets derive from the key.

Reply:
Fixed in https://github.com/asgardeo/thunder/pull/85/commits/9cf89ba41c8877121fa7978fbb49e839a312443a